nodeJS/express/using_postgresql.md

```diff
@@ -0,0 +1,309 @@
+
+Client is an individual connection to the DB, which you manually manage. You open a connection, do your query, then close it. This is fine for one-off queries, but can become expensive if you're dealing with a lot of queries. Wouldn't this problem be alleviated if we could somehow hold onto a client? Yes!
+
+Enter pool. As the name suggests, it's a pool of clients. A pool holds onto connections. And when you query, it'll programmatically open a new connection unless there's an existing spare one. Perfect for web servers.
+
+</div>
+
+### Querying with pg
+
+With our initialized `Pool`, we can use the `query` method. Create a new `db/queries.js` file. Upon revising our project requirements, we understand we need two db interactions: getting all usernames and inserting a new username. Let's define these functions:
+
+```javascript
+const pool = require("./pool");
+
+async function getAllUsernames() {
+  const { rows } = await db.query("SELECT * FROM usernames");
+  return rows;
+}
+
+async function insertUsername(username) {
+  await db.query("INSERT INTO usernames (username) VALUES ($1)", [username]);
+}
+
+module.exports = {
+  getAllUsernames,
+  insertUsername
+}
+```
+
+<div class="lesson-note" markdown="1">
+
+#### Parameterization
+
+What's with the `$1` in the insert query?
+
+Alternatively, the query could look like:
+
+```javascript
+await db.query("INSERT INTO usernames (username) VALUES ('" + username + "')");
+```
+
+We're passing user entered value i.e. `username` directly into our query. A nefarious user could enter something like `sike'); DROP TABLE usernames; --` and wreak havoc. Scary stuff. This is called [SQL injection](https://en.wikipedia.org/wiki/SQL_injection).
+
+`pg` provides [query parameterization](https://node-postgres.com/features/queries#parameterized-query) to prevent this. Instead of passing user input directly, we pass it in an array as the second argument. `pg` handles the rest.
+
+</div>
+
+Invoke the above two functions in the specific controllers (you might have different function names etc. Important thing is to understand how the db functions are invoked):
+
+```javascript
+const db = require("../db/queries");
+
+async function getUsernames(req, res) {
+  const usernames = await db.getAllUsernames();
+  console.log("Usernames: ", usernames);
+  res.send("Usernames: " + usernames.map(user => user.username).join(", "));
+}
+
+async function createUsernameGet(req, res) {
+  // render the form
+}
+
+async function createUsernamePost(req, res) {
+  const { username } = req.body;
+  await db.insertUsername(username);
+  res.redirect("/");
+}
+
+module.exports {
+  getUsernames,
+  createUsername
+}
+```
+
+Take your app for a spin, hopefully it works as expected.
+
+### Populate the db via a script
+
+You might have noticed how cumbersome it is to create a table and populate it with data. Luckily, we have the power of c(n)ode by our side, let's automate it via a script. Create a new file `db/populatedb.js`.
```

Comment:
Should we also explicitly say here about replacing the connection string for the Client as necessary?

Side note: `c(n)ode` lmao :joy:

Reply:
> Should we also explicitly say here about replacing the connection string for the Client as necessary?

^ not sure what you mean

Reply:
Referring to the Client connection string in the example below
```js
  const client = new Client({
    connectionString: "postgresql://<role_name>:<role_password>@localhost:5432/top_users",
  });
  ```
Not a big thing really, only wondering if it wouldn't hurt to add a little mention either here or as a comment in the example like 
`// replace with your own connection string`
or something

Reply:
they shall experience the pain of blindly copy pasting stuff 🙈